---
title: Research
layout: page
permalink: /research/
# image: assets/images/research.png
---

Broadly, I'm interested in identifying the molecular mechanisms underlying the production of phenotypic variation in phenotypic traits. During my Ph.D. and postdoc with [Todd Castoe](https://www.castoelaboratory.org/), my focus was investigating the molecular drivers of phenotypic variation in the expression of different snake venom genes from venom glands. Not only are we studying why venoms are so variable, but also much more generally how different phenotypes arise by fundamentally different genomic processes. In the lab, we leverage a broad array of functional genomic methods to study snake venom variation like single-cell sequencing methods, Hi-C, ATAC-seq, RNA-seq, ChIP-seq and whole genome sequencing. Below, I go into more detail on specific research themes and work I've been involved with.

<br>

## Divergent gene regulation and phenotypic variation in non-model systems

![left-aligned-image]({{ site.url }}{{ site.baseurl }}/assets/images/research1.png){: .align-left}
A unifying theme across many fields of biology is understanding how changes in genotype cause changes in phenotype. My doctoral work was centered on this topic. My first contribution was to a study that identified many of the genomic sequences and transcription factors that control the production of rattlesnake venom [(Perry et al. 2022)](https://www.genome.org/cgi/doi/10.1101/gr.276251.121). This directly led to supplemental work characterizing the genomic structure of major venom component in rattlesnakes [(Gopalan et al., 2022)](https://doi.org/10.1016/j.toxicon.2022.06.014) and understanding how regulatory network variation across a phylogenetic sampling drove divergent venom expression profiles [(Gopalan et al. 2024)](https://doi.org/10.1093/gbe/evae110).

<br>

## Using single-cell methods to detect variation-producing mechanisms

Single-cell sequencing methods are a relatively new technique that we've used to identify molecular pathways that produce tissue-level expression heterogeneity [(Westfall et al., 2022)](https://doi.org/10.1093/gbe/evad109). 

As a hypothesis testing platform, the predictive power of single-cell sequencing is affected by the number of cells sampled and the types of data that can be measured at once. Future directions for this work are to leverage the combined measurement of gene expression and chromatin state to precisely measure how changing genome accessiblity affects gene expression. This type of "multi-omic" assay provides an incredible increase in power to detect cis-regulatory elements compared to other tissue-based methods that will provide very detailed maps of regulatory interactions in the genome.
