---
title: Research
layout: page
permalink: /research/
image: assets/images/research.png
---

In the broadest terms, my interests lie in understanding the molecular mechanisms underlying the production of phenotypic variation in complex traits. During my Ph.D. with [Todd Castoe](https://www.castoelaboratory.org/), my focus was on investigating the production of phenotypic differences in snake venom systems. Snake venoms are notable for the incredible variation in venom composition, even between individuals of the same species. Delving into this topic required the combined use of several different genomic techniques, such as RNA-seq, ATAC-seq, single-cell methods, whole genome sequencing data and others. Below, I go into more detail on specific research themes and work I've been involved with.

<br>

## Linking gene regulatory network divergence to phenotypes

![left-aligned-image]({{ site.url }}{{ site.baseurl }}/assets/images/research1.png){: .align-left}
A unifying theme across biology is understanding how changes in genotype cause changes in phenotype. My doctoral work was primarily centered on this topic. My first contribution was to a study that identified many of the genomic sequences that control the production of rattlesnake venom [(Perry et al. 2022)](https://www.genome.org/cgi/doi/10.1101/gr.276251.121). This directly led to other work examining the genomic structure of an unassembled, but major venom component [(Gopalan et al., 2022)](https://doi.org/10.1016/j.toxicon.2022.06.014) and regulatory network variation across a phylogenetic sampling [(Gopalan et al. 2024)](https://doi.org/10.1093/gbe/evae110). Continuing work in this theme will involve collection of large amounts of data to identify, with high precision, specific molecular changes that modify venom phenotypes. 

<br>

## Using single-cell methods to detect variation-producing mechanisms

Single-cell sequencing methods are a relatively new technique that we've used to identify molecular pathways that produce tissue-level expression heterogeneity [(Westfall and Gopalan et al., 2022)](https://doi.org/10.1093/gbe/evad109). A next step will be to leverage the combined measurement of gene expression and chromatin state to determine how changing genome accessiblity affects gene expression. This type of "multi-omic" assay provides an incredible increase in power to detect cis-regulatory elements compared to other tissue-based methods.
